cpp-package/include/mxnet-cpp/optimizer.hpp

```diff
@@ -125,13 +125,16 @@ inline float Optimizer::GetWD_(int index) {
 }
 
 inline Optimizer* OptimizerRegistry::Find(const std::string& name) {
-  MXNETCPP_REGISTER_OPTIMIZER(sgd, SGDOptimizer);
-  MXNETCPP_REGISTER_OPTIMIZER(ccsgd, SGDOptimizer);  // For backward compatibility
-  MXNETCPP_REGISTER_OPTIMIZER(rmsprop, RMSPropOptimizer);
-  MXNETCPP_REGISTER_OPTIMIZER(adam, AdamOptimizer);
-  MXNETCPP_REGISTER_OPTIMIZER(adagrad, AdaGradOptimizer);
-  MXNETCPP_REGISTER_OPTIMIZER(adadelta, AdaDeltaOptimizer);
-  MXNETCPP_REGISTER_OPTIMIZER(signum, SignumOptimizer);
+  if (cmap().empty()) {
```

Comment:
You are checking all optimizers at once, but wouldn't it be better to check every optimizer separately? AFAICT Registry::__REGISTER__ is used, but Registry::__REGISTER_OR_GET__ would make more sense.

Reply:
Optimizers are all registered at once in the original CPP-Package code and in the Python code using decorators. In the original code [#5251](https://github.com/apache/incubator-mxnet/blob/21b2da07f6ff10669b7649860b913e17e6184708/cpp-package/include/mxnet-cpp/optimizer.hpp), they were all registered at once in the file, so when the program was initialized, all optimizers would already be registered (which is the case in Python), not in the first call to the function Find, but that caused multiple definition errors, so in #5545, the were all registered at once in the first call to Find, which caused the inability to have more than one optimizer bc they would be re-registered with each call, so this change makes sure they are registered once. Why check every optimizer separately when they should all be registered at the same time and only once?

Reply:
I'm not exactly sure what you mean, but it sounds like you want to revert to the original code, but that would break #5545.  @sifmelcara why in #5545 did you
"3. In optimizer.hpp, I delayed the registration of sgd optimizer to the first call of OptimizerRegistry::Find."?